pandas/core/reshape/merge.py

```diff
@@ -906,13 +906,20 @@ def _maybe_coerce_merge_keys(self):
                 continue
 
             # if we are numeric, then allow differing
-            # kinds to proceed, eg. int64 and int8
+            # kinds to proceed, eg. int64 and int8, int and float
             # further if we are object, but we infer to
             # the same, then proceed
             if is_numeric_dtype(lk) and is_numeric_dtype(rk):
                 if lk.dtype.kind == rk.dtype.kind:
                     continue
 
```

Comment:
also, let's change this to an ``if/elsif/else`` clause (and use pass rather than continue to delineate the cases)